### ISSUE TITLE
First fast boot attempt: ELF starts executing before pnach loads

### ISSUE THREAD

**Issue by Aced14** [PCSX2/pcsx2#627]
**PCSX2 version:**
1.3.1 GIT 2015-04-27 (g0ab0c6c)

**PCSX2 options:**
Cheats or widescreen patches

**Plugins used:**
GSdx, LyliPad-Scp-r5875, SPU2-X, cdvdGigaherz, nuvee ps2 usb 0.2.0, FWnull Driver and GigaHeRz's DEV9 Driver

**Plugin settings:**
GSdx software mode and LyliPad-Scp's DS3 native mode; no settings modifications affect this issue

**Description:**
I recently created updated widescreen hacks for Smuggler's Run (not yet released). One of these was an ELF hack meant to replace an old "live" memory hack. Unfortunately my hack only worked intermittently. I'm experienced with creating ELF hacks and nothing seemed wrong with this particular one. Nevertheless, I tried modifying it in several ways in the hopes of getting it working 100% of the time to no avail.

Upon further experimentation, I found that the only circumstance in which the pnach version of my hack doesn't work is if I fast boot the game right after launching PCSX2 (subsequent fast boots work within the same running instance of PCSX2). PCSX2 starts the game, runs the original MIPS instruction my hack is supposed to override, then replaces the original instruction with my injected one. This makes my hack all but useless, because the hack's instruction isn't getting injected before the point in time it needs to run at.

This issue affects pnaches across-the-board, in every game.

**How to replicate:**

1. Launch PCSX2 and run any game. In the log window, take note of the game's CRC and entrypoint.
2. Create a pnach file named after your game's CRC (example: 0E7F91DA.pnach) and insert this line of code into it:  
``patch=1,EE,????????,extended,0000000D //Break - Replace ???????? with your game's entrypoint and replace the first 0 with a 2 (example: 20100008).``
3. Place your pnach file in either PCSX2's /cheats/ or /cheats_ws/ folder.
4. Exit PCSX2 completely. You shouldn't have any PCSX2 windows open at all.
5. Launch PCSX2, enable cheats or widescreen patches (depending on where you placed your pnach file), then fast boot your game. The game should start normally.
6. In the same running instance of PCSX2, shut down the game, then fast boot it again. The game should immediately freeze at a black screen. This step can be repeated over and over again with the same outcome.
7. Step 5 isn't working properly because PCSX2 isn't injecting the pnach's breakpoint until after the game's ELF entrypoint instruction has already executed. Step 6 works properly.

**Last known version to work:**
Never worked (oldest tested version: R5267).

**PC specifications:**
CPU: Intel Core i7 2600K @ 3.4GHz; GPU: XFX Radeon HD 6970 2GB; OS: Windows 7 SP1

**Other comments:**
N/A

**Comment by gregory38** [comment]
It seems doable

**Comment by gregory38** [comment]
The issue is the function that inject error is executed at run time. It means so intsruction are already compiled and cached.
```
recompilation:
	if (HWADDR(startpc) == ElfEntry)
		xCALL(eeGameStarting);
```
Maybe we can call eeGameStarting directly at the compilation. I will create a branch so it can be tested.

**Comment by gregory38** [comment]
See PR #963

**Comment by gregory38** [comment]
Take 2 on #966

**Comment by Aced14** [comment]
@gregory38 Thanks :)!

**Comment by Aced14** [comment]
@gregory38 Looks like this issue will need to be re-opened :(.

I just got around to downloading take 2's build ([v1.3.1-1452-g07ec921](https://github.com/PCSX2/pcsx2/commit/07ec92175f476b11028eb63050e8a5b4553f2466)) and tried setting entrypoint breakpoints in a couple of games via widescreen pnaches. The games are still starting up normally on my first fast boot attempt with the hacks enabled. Subsequent fast boot attempts enter the breakpoint.

In other words, the outcome of step 5's reproduction steps in the OP haven't changed with #966.

**Comment by gregory38** [comment]
Well I didn't test it. But I was confident :(

**Comment by Aced14** [comment]
The reason this issue doesn't occur when repeatedly fast-booting is likely related to #1310.

The initial fast boot attempt's pnach is continuously written into memory until shortly *after* a second fast boot. That ensures the entrypoint breakpoint is in place prior to execution at the beginning of the second fast boot.

Btw to anyone that stumbles upon this issue wondering why on earth I'm setting an entrypoint breakpoint, see #963's discussion for more context. The underlying cause may potentially have a broad impact on widescreen hacking (particularly ELF hacks).

**Comment by gregory38** [comment]
@Aced14 
one question could you post your patch entry? Did you use patch=1.... or patch=0.... 
Long story short, my hack uses the case 0 not 1.

**Comment by Aced14** [comment]
@gregory38 My sample pnach hack used case 1. Here it is (copied from the OP): ``patch=1,EE,????????,extended,0000000D //Break - Replace ???????? with your game's entrypoint and replace the first 0 with a 2 (example: 20100008).``

I just tried a few tests with the aforementioned hack in cff8cb137cc3f1f73096d54cbecc6d79cd0781ff. Tried one test with case 0 and another with case 1. I used War of the Monsters with cheats enabled. Both tests failed. The break hack never executed in time, even when repeatedly fast booting.

Repeatedly fast booting used to trigger the break in older builds, as a result of the bug identified in #851 and #1310. But @avih fixed it in 56adb85a87c763f078b57f395cdf8fc60917ca32 / 67aee8a19cf9e98bc3fe87a0f2a28eccf40e9329.

**Comment by avih** [comment]
OK, I can see why the commit which applies the pnaches at `recRecompile` doesn't do anything. At the time it's trying to apply the patches, they are still not loaded (apparently that happens very shortly afterwards), so it's applying all loaded patches, but there are none such yet. Regardless though - it only tries to apply place=0 patches.

However, very quickly afterwards the patches do get loaded (not yet applied) and then there's `iR3000A Recompiler reset.` message and then place=0 patches do get applied at the original place they were always applied - at `SysCoreThread::GameStartingInThread()`.

Again, both of these only try to apply the place=0 patches at this stage (and the first doesn't work as noted), so any line which starts with `patch=1, ...` is not yet applied regardless.

The first vsync happens shortly after, and with it the application of the place=1 patches.

Assuming @gregory38 did choose the correct place to patch the memory (and I think he did), the only issue remaining is that we need to have the patches loaded at that stage, and that in order to get applied at that stage, the patch should have place=0 (`patch=0, ...`) (note that this doesn't prevent the pnach file from having the same patch also with place=1 which will also get applied on every vsync, even if specifically it might not be relevant to this case).

Since the patch loading code doesn't really depend on anything (it just fetches data from disk and populates some static data structures), it should be possible to load the patches before trying to apply them at `recRecompile`.

The issue, code-wise, is thread safety. The patch loading happens on the `AppCoreThread`, and the patch application happens at `SysCoreThread`. So we can either load it from the latter while synchronizing with the former, or maybe we can post a "config changed" event and wait until it's handled at the former, and then resume the latter.

It should be possible, but possibly a bit tricky.

**Comment by avih** [comment]
@Aced14 let me know if it's fixed for you. As far as I can tell your test case now works correctly. Please confirm.

**Comment by Aced14** [comment]
@avih In 27e7ecce655bdf4e9984337aae3ef935c1542d1c, ``patch=0`` is now working properly. But ``patch=1`` doesn't seem to.

IMO it seems strange for 0 and 1 to produce different results in this situation. Since 0 is meant to overwrite once and 1 is meant to always overwrite, I'd expect 1 to be able to do anything that 0 can, but on an endless basis.

Is there any chance 1 could be brought up to par with 0 in this kind of scenario? Should this issue be reopened in the meantime? Technically, the OP's reproduction steps are still failing since they were based on 1.

**Comment by avih** [comment]
> IMO it seems strange for 0 and 1 to produce different results in this situation. Since 0 is meant to overwrite once and 1 is meant to always overwrite...

Citation needed ;)

As I explained at https://github.com/PCSX2/pcsx2/issues/851#issuecomment-235732723 , there's no specification to the pnach format (do correct me if I'm wrong, but it won't affect what I say next below). It was invented for PCSX2 by a past PCSX2 dev, and as far as I can tell it's not documented anywhere. Other PCSX2 devs also couldn't find any documentation on it. As it stands, the documentation is the implementation.

I did document it recently at eb1e8902, and note that I did add a comment asking whether or not "1" should mean "0 +1".

I also asked other devs, and while some thought it probably wouldn't harm much to make 1 get applied also at the very beginning where 0 is applied, there was also consensus that we probably shouldn't do it, because right now it's possible to have two lines with the same address where one has 0 and the other has 1, but if 1 meant 0+1 and there are cases where applying a patch too early can hurt, then it would not be possible to override it.

Bottom line, you can make a patch get applied both very early and later continuously with two lines, and the 1 patches were never applied so early, so for the sake of not breaking compatibility with all the existing 1 patches out there, as unlikely as that might be, IMO it should stay as is:
- `patch=0,...` gets applied very early during load.
- `patch=1,...` gets applied continuously as it always have, and typically starts a bit later than where '0' gets applied.

Agreed? :)

**Comment by Aced14** [comment]
@avih I agree with your reasoning :). Although I should note that due to #1482's issue, in the past 1 used to work identically to how 0 now works when full booting (at least in my BIOS, so long as it didn't crash).

Call me crazy, but... do you think there'd be any value in introducing something along the lines of ``patch=2`` to perform both 0+1's functions with a single line?

I've never encountered any situations that I'd consider to warrant 0+1, but it can't hurt to have all options covered. On the other hand, it might not be worth the development effort unless anyone runs into a real-world need for it.

**Comment by avih** [comment]
> On the other hand, it might not be worth the development effort unless anyone runs into a real-world need for it.

It's not an issue of effort, as it would be trivial to either make 1 apply also when 0 is applied or add another value which is the equivalent of 0+1.

However, the definition of pnach patches (as defined by the implementation, but still) has been around for ages, and there are tons of patches out there which I really don't want authors to reconsider rewriting, or that it would behave differently between new and older versions of PCSX2, etc, especially when the such functionality can, could and did work and still working with such duplication.

Adding a new value could add confusion and FUD, and changing how 1 behaves (applied also with 0) has a rare chance to break things.

Like you said, you haven't encountered cases which warrant 0+1 yet. On the rare cases, if at all, where it's required, a solution is available.

Keeping it as is will not destabilize the ecosystem. I find this to have higher value than adding an option which so far no one needed and even if they do, can still solve.

So, not convinced so far that doing either of those has higher value than not doing them :)
